net/graphsync_fetcher.go

```diff
@@ -210,14 +222,17 @@ func (gsf *GraphSyncFetcher) fetchBlocksRecursively(ctx context.Context, baseCid
 
 	// recursive selector to fetch n sets of parent blocks
 	// starting from block matching base cid:
-	//   - fetch all parent blocks
+	//   - fetch all parent blocks, with messages/receipts
 	//   - with exactly the first parent block, repeat again for its parents
 	//   - continue up to recursion depth
 	selector := gsf.ssb.ExploreRecursive(recursionDepth, gsf.ssb.ExploreFields(func(efsb selectorbuilder.ExploreFieldsSpecBuilder) {
-		efsb.Insert("messages", gsf.ssb.Matcher())
-		efsb.Insert("messageReceipts", gsf.ssb.Matcher())
 		efsb.Insert("parents", gsf.ssb.ExploreUnion(
-			gsf.ssb.ExploreAll(gsf.ssb.Matcher()),
+			gsf.ssb.ExploreAll(
+				gsf.ssb.ExploreFields(func(efsb selectorbuilder.ExploreFieldsSpecBuilder) {
+					efsb.Insert("messages", gsf.ssb.Matcher())
+					efsb.Insert("messageReceipts", gsf.ssb.Matcher())
+				}),
+			),
```

Comment:
This is the real fix (thanks to @hannahhoward). Most of the rest of the change in this file is additionally verifying that the blocks for the message and receipt lists are actually stored (along with the block headers which were already verified).